src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1023,6 +1023,14 @@ private unsafe string ReplaceCore(string oldValue, string? newValue, CultureInfo
                 index = ci.IndexOf(this, oldValue, startIndex, this.Length - startIndex, options, &matchLength);
                 if (index >= 0)
                 {
+                    // There's the possibility that 'oldValue' has zero collation weight (is an empty string equivalent).
```

Comment:
It would be better to move this check outside the loop for the perf sake. it is enough to check this condition first time only.

Reply:
I couldn't quite figure out how to do this. The alternative would be to do an "is this equivalent to empty?" check at the beginning of the method, but that's a full p/invoke and call into NLS and I think it would vastly outweigh a simple if check in the middle of a loop.

Do you have a recommendation for how to do this?

Reply:
You may do the following:
             index = ci.IndexOf(this, oldValue, startIndex, this.Length - startIndex, options, &matchLength);
             if (matchLength == 0)

            do
            {
                if (index >= 0)
                {
                    ...
                 }

                ....

                index = ci.IndexOf(this, oldValue, startIndex, this.Length - startIndex, options, &matchLength);

            }



---
In reply to: [364889287](https://github.com/dotnet/runtime/pull/1514#discussion_r364889287) [](ancestors = 364889287)

Reply:
My concern is that the call to `IndexOf` is itself a p/invoke, which has significant overhead. Certainly more overhead than a single _if_ check (which the branch predictor will optimize not-taken) in the loop.

Reply:
I am fine if you think it is not worth it.